src/SourceBuild/content/eng/tools/ExclusionsLibrary/ExclusionsHelper.cs

```diff
@@ -0,0 +1,196 @@
+
+namespace ExclusionsLibrary;
+
+public class ExclusionsHelper
+{
+    /// <summary>
+    /// Prefix for lines that import other exclusion files.
+    /// </summary>
+    private const string FileImportPrefix = "import:";
+
+    /// <summary>
+    /// Regex to narrow down the scope of exclusions. If null, all exclusions will be used.
+    /// For instance, setting this to "vstest" will consider 
+    /// "src/vstest/file.txt" but not "src/arcade/file.txt".
+    /// </summary>
+    private readonly Regex? _exclusionRegex;
+
+    /// <summary>
+    /// Storage for exclusions that were used during the test run.
+    /// Used to check if a file matches an exclusion.
+    /// </summary>
+    private readonly ExclusionsStorage _storage = new();
+
+    /// <summary>
+    /// Storage for exclusions that were not used during the test run.
+    /// Used to generate a new baseline file with the exclusions that were used.
+    /// </summary>
+    private readonly ExclusionsStorage _unusedStorage;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="ExclusionsHelper"/> class.
+    /// <param name="exclusionsFilePath">Path to the exclusions file.</param>
+    /// <param name="exclusionRegexString">Optional regex to narrow down the scope of exclusions.</param>
+    /// </summary>
+    public ExclusionsHelper(string exclusionsFilePath, string? exclusionRegexString = null)
+    {
+        _exclusionRegex = string.IsNullOrWhiteSpace(exclusionRegexString) ? null : new Regex(exclusionRegexString);
+        ParseExclusionsFile(exclusionsFilePath);
+        _unusedStorage = new ExclusionsStorage(_storage);
+    }
+
+    /// <summary>
+    /// Checks if a file is excluded.
+    /// <param name="filePath">Path to the file to check.</param>
+    /// <param name="suffix">
+    ///     Suffix to narrow down the scope of exclusions.
+    ///     Patterns without any suffixes (suffix == null) will also
+    ///     be checked if the file is not excluded with the provided suffix.
+    /// </param>
+    /// </summary>
+    public bool IsFileExcluded(string filePath, string? suffix = null) =>
+        CheckAndRemoveIfExcluded(filePath, suffix) || (suffix != null && CheckAndRemoveIfExcluded(filePath, null));
+
+    /// <summary>
+    /// Generates a new baseline file with the exclusions that were used.
+    /// <param name="updatedFileTag">Optional tag to append to the updated file name.</param>
+    /// <param name="additionalLines">Optional additional lines to append to the updated file.</param>
+    /// <param name="targetDirectory">
+    ///     Optional target directory for the updated file.
+    ///     If not provided, files will be written to the
+    ///     same directory as the original files.
+    /// </param>
+    /// </summary>
+    public void GenerateNewBaselineFile(string? updatedFileTag = null, List<string>? additionalLines = null, string? targetDirectory = null)
+    {
+        foreach (string file in _unusedStorage.GetFiles())
```

Comment:
The abstraction that `_unusedStorage` is providing is making things inefficient. This call to `GetFiles` will enumerate the whole dictionary to find the keys. Then later, the `GetExclusion` method is called for each of those files. That will do another dictionary lookup. But since you've already enumerated the whole dictionary, you might as well return the values along with the files at that point. And from that standpoint, there's really no need for a dictionary at all since you're just enumerating the whole set. So really I think that `_unusedStorage` could just be of type `List<(string File, List<Exclusion> Exclusions)>` instead.